Patch from pkgsrc.

Support SunOS-specific GCC behaviour.
Limit GNU ld options correctly.

--- mesonbuild/compilers/compilers.py	2018-09-22 15:22:03.000000000 +0000
+++ mesonbuild/compilers/compilers.py	2018-10-13 16:50:30.281255131 +0000
@@ -179,6 +179,14 @@ arm_buildtype_linker_args = {'plain': []
                              'custom': [],
                              }
 
+sunos_buildtype_linker_args = {'plain': [],
+                                 'debug': [],
+                                 'debugoptimized': [],
+                                 'release': [],
+                                 'minsize': [],
+                                 }
+
+
 msvc_buildtype_linker_args = {'plain': [],
                               'debug': [],
                               'debugoptimized': [],
@@ -1136,7 +1144,7 @@ class Compiler:
             # linked against local libraries will fail to resolve them.
             args.append('-Wl,-z,origin')
         args.append('-Wl,-rpath,' + paths)
-        if get_compiler_is_linuxlike(self):
+        if get_compiler_uses_gnuld(self):
             # Rpaths to use while linking must be absolute. These are not
             # written to the binary. Needed only with GNU ld:
             # https://sourceware.org/bugzilla/show_bug.cgi?id=16936
@@ -1181,6 +1189,7 @@ class CompilerType(enum.Enum):
     GCC_OSX = 1
     GCC_MINGW = 2
     GCC_CYGWIN = 3
+    GCC_SUNOS = 4
 
     CLANG_STANDARD = 10
     CLANG_OSX = 11
@@ -1193,7 +1202,7 @@ class CompilerType(enum.Enum):
 
     @property
     def is_standard_compiler(self):
-        return self.name in ('GCC_STANDARD', 'CLANG_STANDARD', 'ICC_STANDARD')
+        return self.name in ('GCC_STANDARD', 'CLANG_STANDARD', 'ICC_STANDARD', 'GCC_SUNOS')
 
     @property
     def is_osx_compiler(self):
@@ -1250,6 +1250,7 @@ def get_compiler_uses_gnuld(c):
         CompilerType.GCC_STANDARD,
         CompilerType.GCC_MINGW,
         CompilerType.GCC_CYGWIN,
+        CompilerType.GCC_SUNOS,
         CompilerType.CLANG_STANDARD,
         CompilerType.CLANG_MINGW,
         CompilerType.ICC_STANDARD,
@@ -1348,6 +1357,8 @@ class GnuLikeCompiler(abc.ABC):
     def get_buildtype_linker_args(self, buildtype):
         if self.compiler_type.is_osx_compiler:
             return apple_buildtype_linker_args[buildtype]
+        elif self.compiler_type.is_standard_compiler == 4:
+            return sunos_buildtype_linker_args[buildtype]
         return gnulike_buildtype_linker_args[buildtype]
 
     @abc.abstractmethod
